streaming/receiver.c

```diff
@@ -0,0 +1,417 @@
+        u_long err;
+        char buf[256];
+        while ((err = ERR_get_error()) != 0) {
+            ERR_error_string_n(err, buf, sizeof(buf));
+            error("STREAM %s [receive from %s] ssl error: %s", r->hostname, r->client_ip, buf);
+        }
+        return 1;
+    }
+#endif
+    if (!fgets(r->read_buffer, sizeof(r->read_buffer), fp))
+        return 1;
+    r->read_len = strlen(r->read_buffer);
+    return 0;
+}
+
+/* Produce a full line if one exists, statefully return where we start next time.
+ * When we hit the end of the buffer with a partial line move it to the beginning for the next fill.
+ */
+static char *receiver_next_line(struct receiver_state *r, int *pos) {
+    int start = *pos, scan = *pos;
+    if (scan >= r->read_len) {
+        r->read_len = 0;
+        return NULL;
+    }
+    while (scan < r->read_len && r->read_buffer[scan] != '\n')
+        scan++;
+    if (scan < r->read_len && r->read_buffer[scan] == '\n') {
+        *pos = scan+1;
+        r->read_buffer[scan] = 0;
+        return &r->read_buffer[start];
+    }
+    memcpy(r->read_buffer, &r->read_buffer[start], r->read_len - start);
+    r->read_len -= start;
+    return NULL;
+}
+
+
+size_t streaming_parser(struct receiver_state *rpt, struct plugind *cd, FILE *fp) {
+    PARSER_USER_OBJECT *user = callocz(1, sizeof(*user));
+    user->enabled = cd->enabled;
+    user->host = rpt->host;
+    user->opaque = rpt;
+    user->cd = cd;
+    user->trust_durations = 0;
+
+    PARSER *parser = parser_init(rpt->host, user, fp, PARSER_INPUT_SPLIT);
+    parser_add_keyword(parser, "TIMESTAMP", streaming_timestamp);
+
+    if (unlikely(!parser)) {
+        error("Failed to initialize parser");
+        cd->serial_failures++;
+        return 0;
+    }
+
+    parser->plugins_action->begin_action     = &pluginsd_begin_action;
+    parser->plugins_action->flush_action     = &pluginsd_flush_action;
+    parser->plugins_action->end_action       = &pluginsd_end_action;
+    parser->plugins_action->disable_action   = &pluginsd_disable_action;
+    parser->plugins_action->variable_action  = &pluginsd_variable_action;
+    parser->plugins_action->dimension_action = &pluginsd_dimension_action;
+    parser->plugins_action->label_action     = &pluginsd_label_action;
+    parser->plugins_action->overwrite_action = &pluginsd_overwrite_action;
+    parser->plugins_action->chart_action     = &pluginsd_chart_action;
+    parser->plugins_action->set_action       = &pluginsd_set_action;
+
+    user->parser = parser;
+
+    do {
+        if (receiver_read(rpt, fp))
+            break;
+        int pos = 0;
+        char *line;
+        while ((line = receiver_next_line(rpt, &pos)))
+            if (unlikely(netdata_exit || rpt->shutdown || parser_action(parser,  line)))
+                goto done;
+        rpt->last_msg_t = now_realtime_sec();
+    }
+    while(!netdata_exit);
+done:
+    info("PARSER ended");
+    return ((PARSER_USER_OBJECT *) user)->count;
```

Comment:
This will leak memory (user and parser) objects

Reply:
Fixed